providers/keycloak_test.go

```diff
@@ -22,142 +36,202 @@ func testKeycloakProvider(hostname, group string) *KeycloakProvider {
 			ValidateURL:  &url.URL{},
 			Scope:        ""})
 
-	if group != "" {
-		p.SetGroup(group)
-	}
+	if backend != nil {
+		bURL, err := url.Parse(backend.URL)
+		if err != nil {
+			return nil, err
+		}
+		hostname := bURL.Host
 
-	if hostname != "" {
 		updateURL(p.Data().LoginURL, hostname)
 		updateURL(p.Data().RedeemURL, hostname)
 		updateURL(p.Data().ProfileURL, hostname)
 		updateURL(p.Data().ValidateURL, hostname)
 	}
-	return p
-}
 
-func testKeycloakBackend(payload string) *httptest.Server {
-	path := "/api/v3/user"
+	return p, nil
+}
 
+func testKeycloakBackend() *httptest.Server {
```

Comment:
What if this took a `func(http.ResponseWriter, http.Request)` and the test server was spun up and down for each test, that way we can encode the server handling logic with the test cases rather than in this function separately

Reply:
Smart - made the refactor as you suggested 👍